content/en/blog/prerendering-preset-vite.md

```diff
@@ -0,0 +1,351 @@
+
+function App() {
+    return <h1>Hello World!</h1>
+}
+
+-render(<App />, document.getElementById('app'));
++if (typeof window !== 'undefined') {
++	hydrate(<App />, document.getElementById('app'));
++}
+
++export async function prerender(data) {
++    return await ssr(<App {...data} />)
++}
+```
+
+With this set up, you will have an app that prerenders. However, no app is really this simple, so let's look at a couple more complex examples.
+
+### Full API Example
+
+```jsx
+// src/index.jsx
+
+// ...
+
+export async function prerender(data) {
+    const { html, links: discoveredLinks } = ssr(<App />);
+
+    return {
+        html,
+        // Optionally add additional links that should be
+        // prerendered (if they haven't already been -- these will be deduped)
+        links: new Set([...discoveredLinks, '/foo', '/bar']),
+        // Optionally configure and add elements to the `<head>` of
+        // the prerendered HTML document
+        head: {
+            // Sets the "lang" attribute: `<html lang="en">`
+            lang: 'en',
+            // Sets the title for the current page: `<title>My cool page</title>`
+            title: 'My cool page',
+            // Sets any additional elements you want injected into the `<head>`:
+            //   <link rel="stylesheet" href="foo.css">
+            //   <meta property="og:title" content="Social media title">
+            elements: new Set([
+                { type: 'link', props: { rel: 'stylesheet', href: 'foo.css' } },
+                { type: 'meta', props: { property: 'og:title', content: 'Social media title' } }
+            ])
+        }
+    }
+}
+```
+
+### Fetch Content Isomorphically with Suspense-based Fetching
+
+```jsx
+// src/use-fetch.js
+import { useState } from "preact/hooks";
+
+const cache = new Map();
+
+async function load(url) {
+	const res = await fetch(url);
+	if (res.ok) return await res.text();
+	throw new Error(`Failed to fetch ${url}!`);
+}
+
+// Simple suspense-based fetch mechanism with caching
+export function useFetch(url) {
+	const [_, update] = useState({});
+
+	let data = cache.get(url);
+	if (!data) {
+		data = load(url);
+		cache.set(url, data);
+		data.then(
+			(res) => update((data.res = res)),
+			(err) => update((data.err = err)),
+		);
+	}
+
+	if (data.res) return data.res;
+	if (data.err) throw data.err;
+	throw data;
+}
+```
+
+```jsx
+// src/index.jsx
+import { hydrate, prerender as ssr } from 'preact-iso';
+import { useFetch } from './use-fetch.js';
+
+function App() {
+    return (
+	    <div>
+		    <Suspense fallback={<p>Loading...</p>}>
+			    <Article />
+		    </Suspense>
+		</div>
+    );
+}
+
+function Article() {
+	const data = useFetch("/my-local-article.txt");
+	return <p>{data}</p>;
+}
+
+if (typeof window !== 'undefined') {
+	hydrate(<App />, document.getElementById('app'));
+}
+
+export async function prerender(data) {
+    return await ssr(<App {...data} />)
+}
+```
+
+### Using `globalThis` to pass data around
```

Comment:
I like the later example of having a global cache to write to, but I'm a bit unsure about passing other data. I'm worried that this could lead to race conditions given that we also support async rendering in RTS now.

Reply:
There might be a better example in there. This sort of thing is something I've pointed out to people enough times that I think I want *something* there, as it's fairly reasonable to need to pass data back up (and out of the tree). Will have a think.

Shouldn't be any race conditions AFAIK; component tree will have to have fully rendered before the RTS pass completes? That's not to say you can't cause issues, but... shouldn't be anything inherent to the pattern?